File: #244_Shrotest_Word_Distance_II/Solution.cpp
#include <string>
#include <vector>
#include <unordered_map>
using namespace std;

class WordDistance {
private:
	unordered_map<string, vector<int> > dic;
public:
	WordDistance(vector<string> words) {
        for (int i = 0; i < words.size(); i++) {
        	dic[words[i]].push_back(i);
        }
    }

    int shortest(string word1, string word2) {
        vector<int> list1 = dic[word1];
        vector<int> list2 = dic[word2];
        int re = INT_MAX;
        for (int i = 0, j = 0; i < list1.size() && j < list2.size();) {
        	if (list1[i] < list2[j]) {
        		re = min(re, list2[j] - list1[i]);
        		i++;
        	}
        	else {
        		re = min(re, list1[i] - list2[j]);
        		j++;
        	}
        }
        return re;
    }
};

int main(int argc, char* argv[]) {
	string str[] = {"practice", "makes", "perfect", "coding", "makes"};
	vector<string> s(str, str + sizeof(str) / sizeof(*str));
	WordDistance wd(s);
	printf("%d\n", wd.shortest("coding", "makes"));
}

File: #244_Shrotest_Word_Distance_II/Solution.py
# O(n) initialization and O(m + n) queries where m & n are number of time that string appears

class WordDistance(object):
    def __init__(self, words):
    	self.dic = {}
    	self.len = len(words)
    	for i in range(self.len):
    		if words[i] not in self.dic:
    			self.dic[words[i]] = []
    		self.dic[words[i]].append(i)

    def shortest(self, word1, word2):
    	list1 = self.dic[word1]
    	list2 = self.dic[word2]
    	re = self.len
    	i, j = 0, 0
    	while i < len(list1) and j < len(list2):
    		if list1[i] < list2[j]:
    			re = min(re, list2[j] - list1[i])
    			i += 1
    		else:
    			re = min(re, list1[i] - list2[j])
    			j += 1
    	return re

sol = WordDistance(["practice", "makes", "perfect", "coding", "makes"])
print sol.shortest("makes", "coding")